Feature pipeline integration tests (#392)
data61/anonlink-entity-service: docs/devops.rst

Devops
===========

> Azure Pipeline

Reviewer: ```suggestion
Continuous Integration
```

Author: I put Azure pipeline because I did not want to also write the documentation for the other CI we are using, but I was expecting that at one point we should. So I'll put Continuous integration as section, and `Azure DevOps` as sub-section.

--------------

> ``entity-service`` is automatically built and tested using Azure Pipeline

Reviewer: ```suggestion
``anonlink-entity-service`` is automatically built and tested using Azure DevOps
```

Author: :+1:

in the project `Anonlink <https://dev.azure.com/data61/Anonlink>`.

It consists only of a `build pipeline <https://dev.azure.com/data61/Anonlink/_build?definitionId=1>`.

> The build pipeline is described by the script `azurePipeline.yml`

Reviewer: ```suggestion
The build pipeline is defined by the script `azurePipeline.yml`
```

Author: :+1:

> which is using resources from the folder `.azurePipeline`.

Reviewer: ```suggestion
which uses resources from the folder `.azurePipeline`.
```

Author: :+1:

Mainly, it:

- builds and pushes docker images
  - the frontend ``data61/anonlink-nginx``
  - the backend ``data61/anonlink-app``
  - the tutorials ``data61/anonlink-docs-tutorials`` (used to tests the tutorial Python Notebooks)
  - the benchmark ``data61/anonlink-benchmark`` (used to run the benchmark)

> - runs the benchmark using ``docker-compose`` and publishes the artiafct in Azure

Reviewer: ```suggestion
- runs the benchmark using ``docker-compose`` and publishes the results as an artifact in Azure
```

Author: :+1:

- runs the tutorial tests using ``docker-compose`` and publishes the results in Azure
- runs the integration tests by deploying the whole service on ``Kubernetes``, runs the tests and publishes the results in Azure.   

The build pipeline is triggered for every push on every branch. It is not triggered by Pull

> Requests not to build twice the same commit.

Reviewer: ```suggestion
Requests to avoid duplicate testing and building potentially untrusted code.
```

Author: :+1:

The build pipeline requires two environment variables provided by Azure environment:

- `dockerHubId`: username for the pipeline to push images to Data61 dockerhub
- `dockerHubPassword`: password for the corresponding username (this is a secret variable).

> It also requires a secret file (given in the Azure `Library` tab) containing the ``aws`` cluster

Reviewer: ```suggestion
It also requires a secret file (given in the Azure `Library` tab) containing the ``k8s`` cluster
```

Although we happen to be using AWS I don't think this is aws specific.

Author: That's right: this is the kubeconfig containing the credentials to our aws cluster.
